src/tests.c

```diff
@@ -6856,6 +6860,75 @@ void run_ecdsa_edge_cases(void) {
     test_ecdsa_edge_cases();
 }
 
+#ifdef ENABLE_LIBECC_TESTS
+void test_secp256k1_sign_libecc_verify(void){
+    printf("In libecc test\n");
+    /* generate public private key pair */
+    secp256k1_scalar key;
+    unsigned char privkey_buf[32];
+    unsigned char pubkey_buf[65];
+    size_t pubkey_len = sizeof(pubkey_buf);
+    secp256k1_pubkey secp_pubkey;
+
+    random_scalar_order_test(&key);
+    secp256k1_scalar_get_b32(privkey_buf, &key);
+
+    CHECK(secp256k1_ec_pubkey_create(ctx, &secp_pubkey, privkey_buf) == 1);
+    secp256k1_ec_pubkey_serialize(ctx, pubkey_buf, &pubkey_len , &secp_pubkey, SECP256K1_EC_UNCOMPRESSED);
+
+    /* generate random message */
+    secp256k1_scalar msg;
+    unsigned char msg_buf[32];
+
+    random_scalar_order_test(&msg);
+    secp256k1_scalar_get_b32(msg_buf, &msg);
+
+    /* hash and sign with libsecp256k1 */
+    unsigned char to_sign[32];
+    unsigned char sig_buf[64];
+    secp256k1_ecdsa_signature sig;
+    secp256k1_sha256 sha;
+
+    secp256k1_sha256_initialize(&sha);
+    secp256k1_sha256_write(&sha, msg_buf, sizeof(msg_buf));
+    secp256k1_sha256_finalize(&sha, to_sign);
+
+    CHECK(secp256k1_ecdsa_sign(ctx, &sig, to_sign, privkey_buf, NULL, NULL) == 1);
+    CHECK(secp256k1_ecdsa_verify(ctx, &sig, to_sign, &secp_pubkey) == 1);
+
+    secp256k1_ecdsa_signature_serialize_compact(ctx, sig_buf, &sig);
+
+    /* verify with libecc */
+    const char *curve_name = "SECP256K1";
+    const int curve_name_length = strlen(curve_name) + 1;
+    const ec_str_params *str_params;
+    CHECK(ec_get_curve_params_by_name((const uint8_t *)curve_name, curve_name_length, &str_params) == 0);
+
+    ec_params params;
+    CHECK(import_params(&params, str_params) == 0);
+
+    const ec_sig_mapping *sig_mapping;
+    CHECK(get_sig_by_name("ECDSA", &sig_mapping) == 0);
+
+    pubkey_buf += 1; // this is needed because libecc only supports uncompressed points
+    pubkey_len -= 1; // and does not recoginze B0
```

Comment:
After reading https://github.com/ANSSI-FR/libecc/blob/master/src/sig/ec_key.h several times, I concluded that `libecc` supports importing projective or affine coordinates, in either a "structured" or an "unstructured" format. By structured they mean that the buffer includes the curve parameters, the signature algorithm, etc., in addition to the key. So I thought the simplest way to port `libsecp`'s public key to `libecc` is through an unstructured affine buffer.

By B0 I just mean the first byte :). I took this terminology from [here](https://tools.ietf.org/id/draft-jivsov-ecc-compact-05.html). I was puzzled why the public key produced by `libsecp` is 65 bytes long when a key is really two 32-byte values. So I googled a bit and found that "uncompressed keys" have an extra `04` at the beginning.

Reply:
Are you confusing jacobian/projective coordinates with compressed/uncompressed encoding?

* Affine coordinates are tuples (x, y) on the curve (
* Projective coordinates are tuples (X, Y, Z) representing the affine coordinate (X/Z, Y/Z).
* Jacobian coordinates are tuples (X, Y, Z) representing the affine coordinates (X/Z^2, Y/Z^3).

SECG encoding of public keys is (all representing affine coordinates):
* Compressed encoding
  * (0x02 || x) for even y
  * (0x03 || x) for odd y
* Uncompressed encoding
  * (0x04 || x || y)

Reply:
Thanks for succinctly summarizing the different presentations :), but I don't think I'm confusing them. I believe I've used the correct formats/encodings/presentations in the code..? The only mistake I've done is offsetting `pubkey_buf` twice.

Reply:
@real-or-random I think we have a misunderstanding because I did not word my comment correctly. The best wording will probably be "does not recognize the header byte". `libecc` really only supports uncompressed keys, but throws errors at the header byte. It explicitly checks that the public key is exactly 64 bytes long (for curve secp256k1). Does that clear the misunderstanding?

Reply:
@real-or-random For byte-by-byte comparison I'll do that, but when signing with libsecp and verifying with libecc, isn't it more logical to give libecc access only to the public key? I mean, the verifier doesn't normally have the private key, they're supposed to read the public key and the message from the signer then check the signature. Right?

Reply:
> @real-or-random For byte-by-byte comparison I'll do that, but when signing with libsecp and verifying with libecc, isn't it more logical to give libecc access only to the public key? I mean, the verifier doesn't normally have the private key, they're supposed to read the public key and the message from the signer then check the signature. Right?

So yes, in the real world, the signer obviously wouldn't reveal the secret key.

For our purpose of cross-testing however, this is fine. If libecc recomputes the public key, then the cross-test would also cover the computation of the public key from the secret key. So if there's a difference between the two libs in that algorithm, the test would also catch it.

Reply:
Hm, I think if we verify with libsecp256k1 and libecc in `test_compare_libecc_secp256k1`, then we won't need the other two tests anymore.